# 项目：沃尔玛销售数据

## 分析目标

 通过对沃尔玛销售数据进行数据分析，找出**供应与需求之间的匹配关系**，预测未来X个月/年的销售情况。

## 简介

数据集包括沃尔玛的销售数据。管理层认为全国范围内有多家零售门店在库存管理方面存在问题，希望得到供应与需求之间的关系，并得到未来X月/年的销售情况预测数据，  
数据集变量含义如下：  
-  Store：店铺编号
-  Date：销售周
-  Weekly_Sales：店铺在该周的销售额
-  Holiday_Flag：是否为假日周
-  Temperature：销售日的温度
-  Fuel_Price：该地区的燃油成本
-  CPI（消费者物价指数）：消费者物价指数
-  Unemployment：失业率

## 读取数据

In [31]:
import pandas as pd
import numpy as np
from datetime import datetime

In [7]:
orginal_data = pd.read_csv("walmart_stores_data.csv")
orginal_data.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

## 评估数据

这一部分，我将对orginal_data这个DataFrame的数据进行评估。  

评估主要从两个方面进行：结构和内容，即整齐度和干净度。  
数据的结构性问题指不满足“每列是一个变量，每行是一个观察值，每个单元格是一个值”三个标准的数据。  
数据的内容性问题包括缺失数据、重复数据、不一致数据、无效数据等。

### 评估数据整齐度

In [19]:
orginal_data.sample(20)

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
1728     13  30-04-2010    1785823.37             0        48.33       2.941   
1877     14  11-06-2010    2249570.04             0        69.71       2.809   
3893     28  17-09-2010    1159812.35             0        82.45       3.028   
607       5  08-10-2010     290494.85             0        64.99       2.633   
1712     12  05-10-2012     979825.92             0        81.61       3.966   
2451     18  25-06-2010    1220983.17             0        74.04       2.810   
3432     25  05-02-2010     677231.63             0        21.10       2.784   
5378     38  07-10-2011     449516.29             0        70.44       3.827   
1391     10  03-02-2012    1867403.01             0        56.85       3.543   
4466     32  24-09-2010    1067432.10             0        64.19       2.793   
3269     23  15-06-2012    1543667.68             0        64.73       3.683   
1816     13  06-01-2012    1865752.78             0        33.80       3.080   
914       7  04-03-2011     551378.39             0        21.84       3.232   
6405     45  06-04-2012     899479.43             0        49.73       3.996   
80        1  19-08-2011    1530761.43             0        89.94       3.554   
5223     37  15-07-2011     498749.62             0        85.63       3.575   
1880     14  02-07-2010    2334788.42             0        76.61       2.815   
4621     33  17-12-2010     238875.26             0        61.12       3.125   
3203     23  11-03-2011    1042043.55             0        29.36       3.600   
2571     18  12-10-2012    1074079.00             0        50.97       4.029   

             CPI  Unemployment  
1728  126.380567         8.107  
1877  182.431557         8.899  
3893  126.145467        14.180  
607   212.301952         6.768  
1712  131.075667        10.199  
2451  132.497600         9.269  
3432  204.247194         8.187  
5378  129.693800        12.890  
1391  130.349677         7.545  
4466  190.571326         9.017  
3269  138.129533         4.125  
1816  130.157516         6.104  
914   192.011600         8.818  
6405  190.685171         8.567  
80    215.669311         7.962  
5223  214.108365         8.177  
1880  182.318780         8.743  
4621  126.879484         9.265  
3203  133.736143         5.114  
2571  138.911700         8.243

从抽样的输出结果来看，orginal_data的数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”标准。具体来看每行是每个店铺一个销售周的销售数据，每列是销售数据的相关变量，因此不存在结构性问题。

### 评估数据干净度

In [10]:
orginal_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   object 
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), object(1)
memory usage: 402.3+ KB


从数据输出结果来看，共有6435个观察值，没有变量存在缺失。  

但是，Store变量的数据类型应为字符串，Date变量的数据类型应为日期，Holiday_Flag变量的数据类型应为布尔型。

#### 评估重复数据

根据数据变量的含义来看，当店铺编号和销售周均重复的数据会对分析结果造成影响，因此筛选店铺编号和销售周一致的数据。

In [18]:
orginal_data.duplicated(['Store','Date']).sum()

0

从输出结果来看，不存在店铺编号和销售周均重复的数据，即不存在影响分析的重复数据。

#### 评估无效数据

In [20]:
orginal_data.describe()

Store  Weekly_Sales  Holiday_Flag  Temperature   Fuel_Price  \
count  6435.000000  6.435000e+03   6435.000000  6435.000000  6435.000000   
mean     23.000000  1.046965e+06      0.069930    60.663782     3.358607   
std      12.988182  5.643666e+05      0.255049    18.444933     0.459020   
min       1.000000  2.099862e+05      0.000000    -2.060000     2.472000   
25%      12.000000  5.533501e+05      0.000000    47.460000     2.933000   
50%      23.000000  9.607460e+05      0.000000    62.670000     3.445000   
75%      34.000000  1.420159e+06      0.000000    74.940000     3.735000   
max      45.000000  3.818686e+06      1.000000   100.140000     4.468000   

               CPI  Unemployment  
count  6435.000000   6435.000000  
mean    171.578394      7.999151  
std      39.356712      1.875885  
min     126.064000      3.879000  
25%     131.735000      6.891000  
50%     182.616521      7.874000  
75%     212.743293      8.622000  
max     227.232807     14.313000

从输出结果来看，所有数据均在正常范围内，不存在无效数据。

## 清理数据

创建DataFrame cleaned_data进行数据清理工作。

In [22]:
cleaned_data = orginal_data
cleaned_data.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

待处理问题：  
- 将Store变量的数据类型改为字符串
- 将Date变量的数据类型改为日期
- 将Holiday_Flag变量的数据类型改为布尔型

**将Store变量的数据类型改为字符串**

In [27]:
cleaned_data['Store'] = cleaned_data['Store'].astype('str')
cleaned_data.dtypes

Store            object
Date             object
Weekly_Sales    float64
Holiday_Flag      int64
Temperature     float64
Fuel_Price      float64
CPI             float64
Unemployment    float64
dtype: object

**将Date变量的数据类型改为日期**

In [35]:
cleaned_data['Date'] = pd.to_datetime(cleaned_data['Date'],format='%d-%m-%Y')

In [36]:
cleaned_data['Date']

0      2010-02-05
1      2010-02-12
2      2010-02-19
3      2010-02-26
4      2010-03-05
          ...    
6430   2012-09-28
6431   2012-10-05
6432   2012-10-12
6433   2012-10-19
6434   2012-10-26
Name: Date, Length: 6435, dtype: datetime64[ns]

**将Holiday_Flag变量的数据类型改为布尔型**

In [38]:
cleaned_data['Holiday_Flag'] = cleaned_data['Holiday_Flag'].astype("bool")
cleaned_data.dtypes

Store                   object
Date            datetime64[ns]
Weekly_Sales           float64
Holiday_Flag              bool
Temperature            float64
Fuel_Price             float64
CPI                    float64
Unemployment           float64
dtype: object

## 保存清理数据

将完成清理的数据集保存为文件walmart_stores_data_cleaned.csv

In [40]:
cleaned_data.to_csv("walmart_stores_data_cleaned.csv",index=False)

In [41]:
show_data = pd.read_csv("walmart_stores_data_cleaned.csv")
show_data.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  2010-02-05    1643690.90         False        42.31       2.572   
1      1  2010-02-12    1641957.44          True        38.51       2.548   
2      1  2010-02-19    1611968.17         False        39.93       2.514   
3      1  2010-02-26    1409727.59         False        46.63       2.561   
4      1  2010-03-05    1554806.68         False        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

## 整理数据

## 数据可视化

## 总结